test/functional/lua/with_spec.lua

```diff
@@ -0,0 +1,292 @@
+local t = require('test.testutil')
+local n = require('test.functional.testnvim')()
+local Screen = require('test.functional.ui.screen')
+
+local fn = n.fn
+local api = n.api
+local command = n.command
+local eq = t.eq
+local exec_lua = n.exec_lua
+local matches = t.matches
+local pcall_err = t.pcall_err
+
+before_each(function()
+  n.clear()
+end)
+
+describe('vim._with {buf = }', function()
+  it('does not trigger autocmd', function()
+    exec_lua [[
+      local new = vim.api.nvim_create_buf(false, true)
+      vim.api.nvim_create_autocmd( { 'BufEnter', 'BufLeave', 'BufWinEnter', 'BufWinLeave' }, {
+        callback = function() _G.n = (_G.n or 0) + 1 end
+      })
+      vim._with({buf = new}, function()
+      end)
+      assert(_G.n == nil)
+    ]]
+  end)
+
+  it('trigger autocmd if changed within context', function()
+    exec_lua [[
+      local new = vim.api.nvim_create_buf(false, true)
+      vim.api.nvim_create_autocmd( { 'BufEnter', 'BufLeave', 'BufWinEnter', 'BufWinLeave' }, {
+        callback = function() _G.n = (_G.n or 0) + 1 end
+      })
+      vim._with({}, function()
+        vim.api.nvim_set_current_buf(new)
+      end)
+      assert(_G.n ~= nil)
```

Comment:
I'd move this `assert()` inside `vim._with()` callback. To test that events got triggered *exactly* after setting different buffer.

Same with `win` context.